Introduced the Word2VecSynonymFilter

dantuzi opened the pull request:
If you want to expand your query/documents with synonyms in Apache Lucene, you need a predefined file containing the list of terms that share the same semantics.
It's not always easy to find a list of basic synonyms for a language and, even if you find it, this doesn’t necessarily match your contextual domain.
The term "daemon" in the domain of operating system articles is not a synonym of "devil" but it's closer to the term "process".

Word2Vec is a two-layer neural network that takes as input a text and outputs a vector representation for each word in the dictionary.
Two words with similar meanings are identified with two vectors close to each other.

This contribution integrates this technique with the text analysis pipeline. It automatically generates synonyms on the fly from a Word2Vec model generated using the library DL4J.
Please see our presentation at the Berlin Buzzwords conference: https://pretalx.com/bbuzz22/talk/UYZAUX/

#### For the reviewer:
Almost all contribution consists of new code

### Issue
[Word2Vec model to generate synonyms on the fly](https://github.com/apache/lucene/issues/12168)

dantuzi committed:
Introduced the Word2VecSynonymFilter

rmuir commented on lucene/test-framework/src/java/org/apache/lucene/tests/analysis/BaseTokenStreamTestCase.java:
```diff
@@ -221,6 +223,12 @@ public static void assertTokenStreamContents(
       flagsAtt = ts.getAttribute(FlagsAttribute.class);
     }
 
+    BoostAttribute boostAtt = null;
```
BoostAttribute is not for use in tokenstreams. it is for multitermquery.

rmuir replied:
This is wrong, wrong wrong. Remove any use of boostattribute outside of MultiTermQuery. It documents *not to do this*.

rmuir reviewed (changes requested):
The boostattribute must not be used in this way. Its documentation explicitly says in bold not to do this.

It also makes me ask the question, what happens when this filter is run at index-time?

tteofili commented:
will have a look in the next few days :bow: @alessandrobenedetti

dantuzi commented:
@rmuir we did some tests at both query and index time.
We tried to index some documents using the following CustomAnalyzer which includes our Word2VecSynonymFilter and we verified the content of the index using Luke.

```
Analyzer analyzer = CustomAnalyzer.builder()
      .withTokenizer(ClassicTokenizerFactory.NAME)
      .addTokenFilter(Word2VecSynonymFilterFactory.NAME, "model", "wiki-ita-w2v-model.zip")
      .addTokenFilter(FlattenGraphFilterFactory.NAME)
      .addTokenFilter(LowerCaseFilterFactory.NAME)
      .build();
```

Please find some screenshots in our presentation from slide 32
https://www.slideshare.net/SeaseLtd/word2vec-model-to-generate-synonyms-on-the-fly-in-apache-lucenepdf

rmuir commented:
I think you misunderstand the question. What happens to `BoostAttribute` at index-time? absolutely nothing.

dantuzi commented:
@rmuir we spent some time analyzing the code after your comments.
#### Index Time
At index time the BoostAttribute is ignored as you said but, it looks like we "waste" 4 bytes to store a float and an assignment for each term

#### Query Time
At query time, everything works as expected because in this [PR](https://github.com/apache/lucene-solr/pull/357/files), @alessandrobenedetti, @dsmiley, and @romseygeek agreed to use BoostAttribute as a container for a float boost to be used at query time.

That being said, independently of the Javadoc (that doesn't really explain why a public class can not be used), is there any concrete issue with using the BoostAttribute?

mkhludnev commented:
For matter of fact, PositionLengthAttribute is also ignored by indexer.

rmuir commented:
> I'll leave the BoostAttribute discussion for another time as I don't have access to the code right now, but javaDocs or not, it seems extremely suspicious to have a public class that shouldn't be used, it smells bad practice, quite frankly.

The problem is how attributes work internally. that's why fuzzy's internal boostattribute is public. It is really important that this only be used by fuzzyquery, it is only for multitermquery. That's why it is documented as such.

rmuir commented:
Here is what the documentation says:
```
/**
 * Add this {@link Attribute} to a {@link TermsEnum} returned by {@link
 * MultiTermQuery#getTermsEnum(Terms,AttributeSource)} and update the boost on each returned term.
 * This enables to control the boost factor for each matching term in {@link
 * MultiTermQuery#SCORING_BOOLEAN_REWRITE} or {@link TopTermsRewrite} mode. {@link FuzzyQuery} is
 * using this to take the edit distance into account.
 *
 * <p><b>Please note:</b> This attribute is intended to be added only by the TermsEnum to itself in
 * its constructor and consumed by the {@link MultiTermQuery.RewriteMethod}.
 *
 * @lucene.internal
 */
```

I am not going to leave the discussion for another time. This needs to be fixed.

dsmiley commented:
@rmuir Why is it important that BoostAttribute's usage be constrained to only MultiTermQuery?

alessandrobenedetti commented:
> > I'll leave the BoostAttribute discussion for another time as I don't have access to the code right now, but javaDocs or not, it seems extremely suspicious to have a public class that shouldn't be used, it smells bad practice, quite frankly.
> 
> The problem is how attributes work internally. that's why fuzzy's internal boostattribute is public. It is really important that this only be used by fuzzyquery, it is only for multitermquery. That's why it is documented as such.

>>I am not going to leave the discussion for another time. This needs to be fixed.

Hi @rmuir, first of all, I deeply appreciate the time you are taking to help us on this issue, thank you for that.
When I said "'ll leave the BoostAttribute discussion for another time" I meant today (I was on holiday last week and I had the chance to look at the code only today with @dantuzi  and @ilariapet .
I appreciate the javaDoc comments, but as also @dsmiley mentioned, I would like to understand practical reasons to avoid the use of the BoostAttribute in other areas.
It seems to me just a simple wrapper of a float value.
Having a JavaDoc comment to restrict the usage of a public class is not ideal, but if you can elaborate we are more than happy to improve the contribution.

alessandrobenedetti commented:
Just to let you know that if we don't get any concrete motivation on why the BoostAttribute should not be used, I'll proceed with an in-depth review and merge in the next few days.
Any additional review and comment is hugely welcome!

alessandrobenedetti commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Dl4jModelReader.java:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @lucene.experimental
+ */
+public class Dl4jModelReader implements Word2VecModelReader {
```
For the sake of context, I worked with @dantuzi on deciding if implementing a custom reader(this one) or using dl4j as an imported library.
Multiple attempts were done to include dl4j as a library in Lucene, but the effort and impact was not worth it so we reverted to a simple custom reader class.

There are downsides for this of course, ma it's much more lightweight

msokolov replied:
+1 to avoid dependencies whenever practical

alessandrobenedetti commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecModel.java:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  @Override
+  public RandomAccessVectorValues<float[]> copy() throws IOException {
```
if this copy is meant to do a deep copy, I suspect we'll need to handle it differently, I am not sure it's copying the internal elements, but reusing them?
So a copy could end up adding elements to data structures used by the original Object?

msokolov replied:
It does not need to do a deep copy; the purpose of this method is to enable multiple concurrent accesses to the underlying data. Since this implementation doesn't have any temporary variable into which vectors are decoded (which could be overwritten), I think it's safe to simply `return this`.

alessandrobenedetti reviewed (approved):
The contribution is clean, simple to read, and very well-tested.
I left some minor considerations, that shoud be addressed.

I'll wait for some additional review and potential elaboration on the BoostAttribute and then merge.
This is just a first step for a new feature, so some future iterations are expected (for example I foresee the optional use of similarity as boost, the support for multi term synonyms, other models, ect

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Dl4jModelReader.java:
```diff
@@ -0,0 +1,122 @@
+import org.apache.lucene.util.TermAndVector;
+
+/**
+ * Word2VecModelReader is a Word2VecModelReader that reads the file generated by the library
```
Should this say "Dl4jModelReader is a Word2Vec..." ?

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Dl4jModelReader.java:
```diff
@@ -0,0 +1,122 @@
+        int vectorDimension = Integer.parseInt(headerValues[1]);
+
+        Word2VecModel model = new Word2VecModel(dictionarySize, vectorDimension);
+        reader
```
to me this would read much clearer using a traditional `while` loop:

```
while ((line = reader.readLine()) != null) ...
```

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Dl4jModelReader.java:
```diff
@@ -0,0 +1,122 @@
+
+  private static final String MODEL_FILE_NAME_PREFIX = "syn0";
+
+  private final String word2vecModelFilePath;
```
why pass in the path when it is only used in `toString()`? Can we choose between accepting a `java.nio.Path` that does its own open/close of the zip file and an (anonymous) `InputStream`?

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Dl4jModelReader.java:
```diff
@@ -0,0 +1,122 @@
+        return model;
+      }
+    }
+    throw new UnsupportedEncodingException(
```
I think this exception is really intended for use with character set encodings only. Maybe IllegalArgumentException would fit better?

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Dl4jModelReader.java:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  static BytesRef decodeTerm(String term) {
+    if (term.toLowerCase(Locale.ROOT).startsWith("b64:")) {
```
It seems wasteful to lower case every term here, even when they are not b64-encoded. Also: is this something that dl4j will do consistently throughout the file? If so, we can peek at the first term and then assume the remainder will also be b64-encoded. I also wonder about the `trim()` - why do we need it? Does `Base64.decode` leave garbage at the end of the terms sometimes?

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecModel.java:
```diff
@@ -0,0 +1,93 @@
+  private final int dictionarySize;
+  private final int vectorDimension;
+  private final TermAndVector[] data;
+  private final Map<BytesRef, TermAndVector> word2Vec;
```
did you consider using BytesRefHash?

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecModelReader.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @lucene.experimental
+ */
+public interface Word2VecModelReader extends Closeable {
```
Do we have more than one implementation? I don't think this interface is necessary. Later we can always add it if we have multiple implementations and need to abstract. For now it's just extra stuff to maintain

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecSynonymFilter.java:
```diff
@@ -0,0 +1,111 @@
+      TermAndBoost synonym = synonymBuffer.pollFirst();
+      clearAttributes();
+      restoreState(this.lastState);
+      termAtt.setEmpty();
```
this would be cleaner if we used `BytesTermAttribute`

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/SynonymProvider.java:
```diff
@@ -0,0 +1,42 @@
+   * SynonymProvider constructor
+   *
+   * @param term we want to find the synonyms
+   * @param maxSynonymsPerTerm maximum number of result returned by the synonym search
```
I don't see that we need this interface if its only use is in this one `Dl4jWord2VecSynonymFilter`. Can't we simply refer directly to the implementing class?

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecModel.java:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  public BytesRef binaryValue(int targetOrd) throws IOException {
+    return data[targetOrd].getTerm();
```
This is incorrect - the purpose of this method is to return some bytes representing the vector value. I think instead you ought to simply throw `UnsupportedOperationException` since this implementation will never be used in an indexing context where this method is required.

msokolov commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecSynonymFilter.java:
```diff
@@ -0,0 +1,111 @@
+import org.apache.lucene.util.TermAndBoost;
+
+/**
+ * Applies single-token synonyms from a Word2Vec trained network to an incoming {@link TokenStream}.
```
I don't see any reference to word2vec in here? We just use this abstract SynonymFilter that could conceivably be implemented using a thesaurus. That makes this more powerful than I think we are claiming it to be. Again, let's simply use the concrete classes throughout and call the classes  `Dl4jWord2VecXXX`

msokolov reviewed (commented):
Thanks for this nice contribution! I think it needs a bit of work, and it would be helpful to explain the benefit somewhere. ~~It's possible I missed some discussion about that sorry~~. Update: OK I see you have a summary in this PR :) I added some suggestions how to tighten up some implementations and a recommendation to eliminate some of the abstraction layers.  I am traveling and might take a few days to respond -- please be patient, sorry!

rmuir commented:
> Hi @rmuir, first of all, I deeply appreciate the time you are taking to help us on this issue, thank you for that.
When I said "'ll leave the BoostAttribute discussion for another time" I meant today (I was on holiday last week and I had the chance to look at the code only today with @dantuzi and @ilariapet .
I appreciate the javaDoc comments, but as also @dsmiley mentioned, I would like to understand practical reasons to avoid the usage of the BoostAttribute in other areas.
It seems to me just a simple wrapper of a float value.
Having a JavaDoc comment to restrict the usage of a public class is not ideal, but if you can elaborate we are more than happy to improve the contribution.

The class is public due to the way attributes are implemented. AttributesImpl classes must be public, there's no way to make them package private! Does this make sense? It is in the search package for the fuzzy query to use for this reason, with lots of warnings in its documentation about the usage.

The attribute is needed to manage the priority queue of fuzzy query's rewrite method. Because of this, there are some odd things about its behavior (e.g. allowing negative boosts). It is really not suitable for usage anywhere else in the system.

please address my concern and don't ignore it. i'm -1 on this PR being merged.

rmuir commented:
> Hi @rmuir, we definitely don't want to ignore improvement recommendations, rest assured.
> Sorry if I am pedantic, I just want to understand why it shouldn't be used for additional things(on top of what it was originally designed to).
> Do you have more info on the "odd things about its behavior"?
> Allowing negative numbers shouldn't be an issue, the vector similarity score is 0<=x<=1.
> Thank you again for your time, it's genuinely much appreciated

The class documents that it should not be used for this purpose. I honestly feel the documentation is good: the problem is you didn't read it. Please don't make me repeat myself here.

To be clear, i'm -1 on the issue and this constitutes a veto. it can't be ignored, just fix the issue instead of merging the PR.

rmuir commented:
> Allowing negative numbers shouldn't be an issue, the vector similarity score is 0<=x<=1.

You don't understand. Some day, someone may want to add such safety as a check to the attribute, but they won't be able to because fuzzy query relies on such things. That's a great big reason this attribute should not be used for other purposes!

It was created to be applied to `TermsEnum` and not a tokenstream at all. By using an attribute on `TermsEnum`, we are able to handle this corner-case needed to accelerate top-N selection of fuzzyquery, without adding unnecessary bloat to `TermsEnum` itself.

romseygeek commented:
Would your worries be assuaged if we created a separate `QueryBoostAttribute` class @rmuir?  Then `QueryBuilder` can use that, and we can add checks for negative boosts, javadocs that say it should only be used for query time analysis, etc.

rmuir commented:
> Would your worries be assuaged if we created a separate `QueryBoostAttribute` class @rmuir? Then `QueryBuilder` can use that, and we can add checks for negative boosts, javadocs that say it should only be used for query time analysis, etc.

It would certainly be a LOT less offensive than using FuzzyQuery's BoostAtt?

But I still don't think analysis pipeline should be involved in this way, in query formulation. It is confusing in many ways, especially with boosts that will get silently dropped at index-time, confusing users. So I'd really rather us avoid analysis chain doing this stuff. it is mixing separate concerns in a very bad way: analysis chain is supposed to be about tokenization and boosts are supposed to be in the domain of higher level query parsers/MLT as it represents query intent (and also makes no sense at index time). See my comment above about whether this should even be an analyzer :) 

I really see it closer to "query expansion" (e.g. closer in functionality to MLT).

alessandrobenedetti commented:
Ok, let me rephrase my question then:
Let's assume:
- we don't care why that class was originally created and its JavaDocs comments
- we don't intend to modify it

Quoting @dsmiley "Why is it important that BoostAttribute's usage be constrained to only MultiTermQuery?" (from the code is just containing a float payload).
What makes the BoostAttribute different from the org.apache.lucene.analysis.tokenattributes.PayloadAttribute for example? (aside from the different types of payload BytesRef vs float).
Let's just give a constructive explanation if you know, rather than citing again that JavaDoc :)

N.B. I am not a strenuous defender of using that class, to be clear, I want to do best in terms of readability and quality for the project, I just want my review to improve the work of the original contributors of this patch!

rmuir commented:
i explained it to you multiple times on this issue and i feel the javadoc explanation is already good. Please read it!

rmuir commented:
i'm -1 for this PR. analyzers shouldnt be mixing with query boosts at all. it is a mixing of concerns that we should avoid: please refactor to be something other than an analyzer.

mkhludnev commented:
> i'm -1 for this PR. analyzers shouldnt be mixing with query boosts at all. it is a mixing of concerns that we should avoid: please refactor to be something other than an analyzer.

Presumably expansion logic might be moved from [Word2VecSynonymFilter.java](https://github.com/apache/lucene/pull/12169/files#diff-c7476adadb37ec05d504af2eb6212ed1eae9a9a7a40da913c3d39fef5d28c8fb) into `QueryBuilder.newSynonymQuery(),.newTermQuery()`. Thus, boosting logic would be moved from analyzer into query parsing, but how to make it easily configurable and pluggable into all query parsers for may users? On top of my head ComplexQP probably can configure such pipeline. But how to turn it on&off for classic, Elastic, Solr, etc?

mkhludnev commented:
> I think you misunderstand the question. What happens to `BoostAttribute` at index-time? absolutely nothing.

Literally, BoostAttribute might be handled with something like BoostAsPayloadTokenFilter. Also, I think such synonym expansion might better work in index time, since it can inject context aware synonyms. WDYT?

dantuzi committed:
Merge branch 'apache:main' into word2vec-pr

dantuzi committed:
Removed BoostAttribute and single-implementation interfaces

First iteration addressing comments in the PR
https://github.com/apache/lucene/pull/12169

dantuzi reviewed (commented):
I'm working on a new version where
- I removed the BoostAttribute
- I removed the single-implementation interfaces
- Addressed the other comments

alessandrobenedetti committed:
Dl4j exception message addressing review #12169 comment by msokolov

alessandrobenedetti committed:
improved code readability addressing #12169 PR comments

alessandrobenedetti commented:
I just finished an additional review, and all the most important comments collected in the last few months have been addressed.
There are a couple of minors that can be added in the future by the community if deemed necessary.

We have long way exceeded the time allowed by our company to work on this contribution, unless any serious blockers I plan to merge it tomorrow morning UK time.
Thanks to everyone involved!

alessandrobenedetti committed:
merge in progress - compilation fixes - exactSearch to DO

romseygeek commented:
Hey @alessandrobenedetti , this is causing failures in TestRandomChains:

```
./gradlew :lucene:analysis.tests:test --tests "org.apache.lucene.analysis.tests.TestRandomChains.testRandomChains" -Ptests.jvms=12 "-Ptests.jvmargs=-XX:TieredStopAtLevel=1 -XX:+UseParallelGC -XX:ActiveProcessorCount=1" -Ptests.seed=108D7DCA5790BA75 -Ptests.gui=true -Ptests.file.encoding=US-ASCII
```

You might need to exclude it from the constructors list?

alessandrobenedetti commented:
Ok, I'm holding off the cherry pick to 9.x until Daniele fixes the issue.

Keep you updated
--------------------------
*Alessandro Benedetti*
Director @ Sease Ltd.
*Apache Lucene/Solr Committer*
*Apache Solr PMC Member*

e-mail: ***@***.***


*Sease* - Information Retrieval Applied
Consulting | Training | Open Source

Website: Sease.io <http://sease.io/>
LinkedIn <https://linkedin.com/company/sease-ltd> | Twitter
<https://twitter.com/seaseltd> | Youtube
<https://www.youtube.com/channel/UCDx86ZKLYNpI3gzMercM7BQ> | Github
<https://github.com/seaseltd>


On Mon, 24 Apr 2023 at 15:42, Daniele Antuzi ***@***.***>
wrote:

> @romseygeek <https://github.com/romseygeek> all tests were passing before
> the merge, I'm investigating to understand why we have this issue and fix it
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/lucene/pull/12169#issuecomment-1520185075>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD5JK7Z66HIP6FWIRBLYWDXCZ7LXANCNFSM6AAAAAAVGTZEWE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dantuzi commented:
@romseygeek I can replicate the issue you mentioned only if I explicitly run the string you posted above with the exact `-Ptests.seed` parameter (in all the other cases it works).
I'm trying to understand how you found the bug and who trigger this specific test case, is it run as an automatic test?

romseygeek commented:
Ah, welcome to the world of randomized tests :)  We have automated tests in lucene that use randomized parameters to generate some of their test configurations, as a way of increasing coverage.  This particular one will generate an analysis chain from a random set of token filters and check that it produces tokens that conform to certain behaviours.  This one was picked up by the elasticsearch CI system, but we have ones running on apache hardware as well.  The randomization is controlled by a seed (which is why the failure only reproduces with the specific command line I reproduced above).  Lots and lots of details can be found in the `help/tests.txt` file file.

jimczi commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecSynonymProvider.java:
```diff
@@ -0,0 +1,104 @@
+              word2VecModel,
+              VECTOR_ENCODING,
+              SIMILARITY_FUNCTION,
+              hnswGraph,
```
From what I understand this provider is used by multiple threads to get the expanded terms. However the `HnswGraph` is not usable by multiple threads so this won't work when multiple queries are run in parallel. 
I am not even sure how this would break but I am quite sure that it won't be easy to debug ;).
I don't think the `OnHeapHnswGraph` was meant to be used with multiple threads, the API is stateful (seek, nextNeighbor, ...) so I don't see how it could be used for this purpose.
Note that the `OffHeapHnswGraph` is re-created for each request and that's why the multi-threaded knn search works, we don't have this mechanism for the on-heap version.

jimczi replied:
An easy fix here could be to add a clone method to`HnswGraph` and make it clear that it needs to be called for concurrent access? cc @msokolov

jimczi reviewed (commented):
That's a nice addition to Lucene, thanks @alessandrobenedetti .
I left one comment regarding the multi-thread aspect. I don't think it works currently and that could lead to weird behaviour/bugs. The `OnHeapHnswGraph` does not handle concurrency so it's not possible to use a single instance with multiple threads. It would require some changes in the APIs to make it possible unless I missed something?

dantuzi commented:
@romseygeek, @alessandrobenedetti and me found the bug and we are fixing it right now

jimczi commented:
>  for the time being, we don't have enough capacity to work again on this contribution

It's not a suggestion it's a bug ;). As it is my suggestion would be to revert this change and to work on a fix. If I am right, this filter is not usable at all and would break in weird ways so please don't ignore it.

msokolov commented:
Thanks for calling this out @jimczi. Yeah, that's true -- the javadoc for `HnswGraph` is  incorrect. It says " The graph may be searched by multiple threads concurrently, but updates are not thread-safe." but it this class doesn't actually provide any searching methods, only some methods for iteration that modify internal state, which are clearly not thread-safe.

We might be able to clean up the API to make it clearer what's going on; multiple people seem to be getting confused by this. But I agree this change as it stands is not usable.

Adding a clone() / copy() method to HnswGraph seems like a fairly unobtrusive way to support this. Javadoc definitely needs to be clearer.

alessandrobenedetti commented:
Hi @jimczi , @msokolov , I don't think it was @dantuzi intention to ignore you, I think he just misread your comment as a suggestion of improvement rather than a blocker.

I had Daniele and Ilaria extensively testing and benchmarking the contribution in the last few months, we left the PR open for various months and we thought it was in a good state to merge (being a new feature, not impacting existing works).

I really appreciate you spending the time to find this problem and we'll take a look in detail in 2 days (Wednesday 26th).
Given that, and not entering the technical merit until the 26th (I don't have the time right now to process your insights), is it worth reverting or we can simply fix it/commit a workaround as soon as we find it in the next few days?

jimczi commented:
> Adding a clone() / copy() method to HnswGraph seems like a fairly unobtrusive way to support this. Javadoc definitely needs to be clearer.

I looked at this and I am a bit reluctant to add this method since the `OnHeapHnswGraph` is mutable. That would be confusing and could lead to the same issue we have in this PR. I'd say that in this current form the `HnswGraph` is an internal class that shouldn't be used outside of the indexing primitives. 

I am also curious to hear what's the expectation for this filter. I checked some resources created with word2vec or GloVe and   they output a large vocabulary size. With a single thread, building a graph of 1M vectors with hundreds of dimensions would take several minutes (16 minutes at 1000 docs/s). That seems overkill for an application to rebuild this graph on every restart. Could we instead build the resource offline (the hnsw graph) and just load it when the filter starts? 

> is it worth reverting or we can simply fix it/commit a workaround as soon as we find it in the next few days?

I don't have a strong feeling here but I am not sure we should add this filter in the common analysis module. As it stands it feels more experimental so the sandbox might be more appropriate?

alessandrobenedetti commented:
I agree @romseygeek and done that, we'll think about the other issues tomorrow on main!

zhaih commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/word2vec/Word2VecSynonymProvider.java:
```diff
@@ -0,0 +1,104 @@
+              SIMILARITY_FUNCTION,
+              hnswGraph,
+              null,
+              word2VecModel.size());
```
setting this limit will actually cause inaccurate result, since this limit is actually limit on "total number of nodes you visited on all levels" so when you reaches the bottom level you're not guaranteed to be able to visit every possible node. (see update of the limit [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java#L171) and reset of `visited` nodes for each level [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java#L223))

alessandrobenedetti replied:
I see, you are referring to line 88 and 'word2VecModel.size()' @dantuzi
